perfdash/google-gcs-downloader.go

```diff
@@ -47,26 +47,26 @@ type GoogleGCSDownloaderOptions struct {
 	OverrideBuildCount bool
 }
 
-// GoogleGCSDownloader that gets data about Google results from the GCS repository.
-type GoogleGCSDownloader struct {
-	GoogleGCSBucketUtils *bucketUtil
-	Options              *GoogleGCSDownloaderOptions
+// Downloader that gets data about Google results from the GCS repository.
+type Downloader struct {
+	BucketUtils BucketUtil
+	Options     *DownloaderOptions
 }
 
-// NewGoogleGCSDownloader creates a new GoogleGCSDownloader.
-func NewGoogleGCSDownloader(opt *GoogleGCSDownloaderOptions) (*GoogleGCSDownloader, error) {
+// NewDownloader creates a new Downloader.
+func NewDownloader(opt *DownloaderOptions) (*Downloader, error) {
 	b, err := newBucketUtil(opt.LogsBucket, opt.LogsPath, opt.CredentialPath)
 	if err != nil {
 		return nil, err
 	}
```

Comment:
This is good, you could probably just make this accept the `BucketUtil` interface and the DownloaderOptions

```suggestion
func NewDownloader(opt *DownloaderOptions, util BucketUtil) *Downloader {
    return &Downloader{
		BucketUtils: util,
		Options:     opt,
	}
}
```

Reply:
Where would we call NewGCSBucketUtil then? Shouldn't NewDownloader create a new bucket util? 

Since we are going to have an s3 flag, I was thinking we can just check if the flag is on, and if so, then do NewS3BucketUtil instead.

Reply:
This gets us into the topic of dependency injection. It's worth a longer discussion in person, but the general idea is to have all of your class instantiation happen in some common initialization area and then wire the classes up like
```
<main.go>
bucketUtil := NewS3BucketUtil()
if (FLAGS.gcs) {
  bucketUtil = NewGcsBucketUtil(..)
}
downloader := new Downloader(options{}, bucketUtil)
foo := new AnotherClassThatNeedsbucketUtil(bucketUtil)
```

The alternative, as you're suggesting, delegates to the inner classes to discover and detect go process flags. While this isn't so bad with a single layer of dependency injection, it becomes a nightmare as you start to have larger and larger DAGs of classes injected into each other. Imagine every class in a huge project making decisions about what instantiation to use. 

What if someone else needs the BucketUtil and that component decides to make an S3BucketUtil instead? How do these communicate on which implementation? What if the name of your flag changes, do you need to change the discovery logic all over the code base?

As with all things, this is a big nuanced subject with no "right answer". In this case, it probably doesn't matter much since it's such a simple hierarchy. However, following these types of conventions can make it much easier for developers to make assumptions about your code and help the project evolve, aka being "extensible".